opm/simulators/linalg/bda/CPR.cpp

```diff
@@ -0,0 +1,515 @@
+                }
+                nnz_idx++;
+            }
+        }
+
+        Opm::BdaBridge<DuneMat, DuneVec, 1>::getSparsityPattern(A, Amatrices.back().rowPointers, Amatrices.back().colIndices);
+
+        // compute inverse diagonal values for current level
+        invDiags.emplace_back(A.N());
+        for (unsigned int row = 0; row < A.N(); ++row) {
+            invDiags.back()[row] = 1 / Amatrices.back().nnzValues[diagIndices[level][row]];
+        }
+    }
+}
+
+
+template <unsigned int block_size>
+void CPR<block_size>::analyzeAggregateMaps() {
+
+    Pmatrices.clear();
+    Rmatrices.clear();
+
+    const DuneAmg::AggregatesMapList& aggregatesMaps = dune_amg->aggregatesMaps();
+
+    DuneAmg::AggregatesMapList::const_iterator mapIter = aggregatesMaps.begin();
+    for(int level = 0; level < num_levels - 1; ++mapIter, ++level) {
+        DuneAmg::AggregatesMap *map = *mapIter;
+
+        // get indices for each row of P and R
+        std::vector<std::set<long int> > indicesP(level_sizes[level]);
+        std::vector<std::set<long int> > indicesR(level_sizes[level+1]);
+        using AggregateIterator = DuneAmg::AggregatesMap::const_iterator;
+        for(AggregateIterator ai = map->begin(); ai != map->end(); ++ai){
+            if (*ai != DuneAmg::AggregatesMap::ISOLATED) {
+                const long int diff = ai - map->begin();
+                indicesP[diff].insert(*ai);
+                indicesR[*ai].insert(diff);
+            }
+        }
+
+        Pmatrices.emplace_back(level_sizes[level], level_sizes[level+1], level_sizes[level]);
+        Rmatrices.emplace_back(level_sizes[level+1], level_sizes[level], level_sizes[level]);
+        std::fill(Pmatrices.back().nnzValues.begin(), Pmatrices.back().nnzValues.end(), 1.0);
+        std::fill(Rmatrices.back().nnzValues.begin(), Rmatrices.back().nnzValues.end(), 1.0);
+
+        // set sparsity pattern of P
+        int col_idx = 0;
+        Pmatrices.back().rowPointers[0] = 0;
+        for (unsigned int i = 0; i < indicesP.size(); ++i) {
+            Pmatrices.back().rowPointers[i + 1] = Pmatrices.back().rowPointers[i] + indicesP[i].size();
+            for (auto it = indicesP[i].begin(); it != indicesP[i].end(); ++it) {
+                Pmatrices.back().colIndices[col_idx] = *it;
+                col_idx++;
+            }
+        }
+        // set sparsity pattern of R
+        col_idx = 0;
+        Rmatrices.back().rowPointers[0] = 0;
+        for (unsigned int i = 0; i < indicesR.size(); ++i) {
+            Rmatrices.back().rowPointers[i + 1] = Rmatrices.back().rowPointers[i] + indicesR[i].size();
+            for (auto it = indicesR[i].begin(); it != indicesR[i].end(); ++it) {
+                Rmatrices.back().colIndices[col_idx] = *it;
+                col_idx++;
+            }
+        }
+    }
+
+}
+
+
+void solve_coarse_umfpack(const Matrix *A, std::vector<double> &y, std::vector<double> &x) {
+    const int N = A->N;
```

Comment:
I do not understand why we are doing it the way we do it here. Here is my understanding of what we do:

- We use the coarsest matrix (not in BCRSMatrix format) despite having the BCRSMatrix lying around anyway
- Because UmfPack-bindings of dune-istl do not understand it we convert it back to BCRSMatrix
- I assume the UmfPack bindings now convert it back to a matrix of C-pointers (but using column compressed row storage).

I would recommend to either use the available BCRSMatrix or to use the C respresentation, use umfpack routines to transpose it to column compressed storage (User guide Section 12) and call umfpack directly.

In any case we should probably only setup umfpack (decompose) if the matrix changes and not in every coarse grid solve.

Reply:
6fa456ea
Now reuses the umfpack object. `setMatrix()` takes about 3.4ms, `apply()` only 0.25ms. This used to be 4.1ms each linear iteration due to the things you mentioned.
I also removed `solve_coarse_umfpack()`, it would now only consist of one line.